### ISSUE TITLE
Android file manager widget/module (4.4 and up)

### ISSUE THREAD

**Issue by Maay** [dialogs/android-filolo#1]
The main purpose of the widget is to let the user select one or more files on the file system, and return control to the host application for the list of files selected by the user (addresses on the file system).

User case (messenger example):

1. User clicks “Share File”

2. The file manager interface opens.

3. The user selects the storage (one of the internal or external storage).

4. Goes to the directory of his interest.

Further two possible scenarios

5.1. Tap on one file transfers control with the address on the device to the host application (messenger), sending
Тап на один файл передает управление с адресом на устройстве приложению-хосту (мессенджеру), send process is initiated

5.2. Long tap switches the multi-selection mode when the user has the ability to select several files, and by clicking on the "Done" button, transfers control to the application.

Required features:

- Navigation in FS with the ability to get to a higher level

- Ability to close the widget without selecting files

Additional features:

- Localization

- Styles customization

- Ability to select multiple files from different directories

- Ability to view a list of files before the final sending to the host application

- Ability to filter files by type/extension

- Different design for phones and tablets

**Comment by mul1sh** [comment]
@Maay  can i work on this one? thanks

**Comment by Maay** [comment]
@mul1sh any luck?

**Comment by mul1sh** [comment]
@Maay  yes i've been working on this will PR tomorrow with the working widget 🙂

**Comment by Maay** [comment]
@mul1sh problems with that PR? 😄

**Comment by mul1sh** [comment]
@Maay  yeah i had some issues but i'm resolving them now 😃 expect the pr in 2 days or less please

**Comment by willy2dg** [comment]
Hi @Maay , i gonna take a look at this if seems okay to you. Just to be sure, this is to be included in other apps or to be installed by the user to by used from other apps ?

**Comment by nkt** [comment]
@willy2dg this is a library to be used in other apps

**Comment by Maay** [comment]
@willy2dg I'd be happy if you try it.

**Comment by willy2dg** [comment]
@Maay @nkt Well, I'll start working on it.

Just some clarifications.

- As localization do you mean string translation files?
- About style customization, should be enought using the values.xml  and similar files to UI modularization ? Which allow you to easyly chage this values? or what kind of customization do you want ?
- Could be enought filtering by extensions ? This filter is for the user or to the app launching the module ? 

I see two posible aproach to implement the extension filter: Show a hardcoded list of extension, which can be big but will persist while navegate directories. Or showing the extensions on current directory, which will look cleanest but can't persist while navigate. What do you prefer ?

**Comment by deminnsk** [comment]
Hi @willy2dg 

> * As localization do you mean string translation files?
Yep, resources localisation is good enough 



> * About style customization, should be enought using the values.xml  and similar files to UI modularization ? Which allow you to easyly chage this values? or what kind of customization do you want ?
Mostly this is about styles: https://developer.android.com/guide/topics/ui/look-and-feel/themes



> * Could be enought filtering by extensions ? This filter is for the user or to the app launching the module ?
Not sure that I clearly got it, but the filter is only related to file system module. And this FM module we should be able to use as gradle dependency for different apps as a library.



> I see two posible aproach ... What do you prefer ?
Better to have both approaches and configure the module from code.

**Comment by willy2dg** [comment]
@Maay @deminnsk generating a dynamic filter seems to be a bad aproach, since it requires many time to inflate all the checkbox. So i recomend to change it for a static view with specified extensions or use a search box to avoid loading times, but i can keep this aproach if you want (its currently implemented), what do you think ?

**Comment by willy2dg** [comment]
I found a good way to generate the dynamic view on activity starts,but creating the filter for every directory will require loading times ( every time you open the filter ) and require reseting the filter every time ( since the extensions will change between directories ), so if you agree i will keep this aproach which can be customiced for every app using the module.

**Comment by nkt** [comment]
@willy2dg performance is important for us, so implement the most performant approach.

**Comment by willy2dg** [comment]
I have had trouble finding a uniform way to implement this module, since the access to storage has changed a lot since version 19 (4.4 KitKat). After much research I have found a way to meet all your requirements since version 21, but with some limitation for versions 19-20 that you should know. I see 2 options for these versiones:

1 - Using a "hacky" way ( parsing file /proc/mounts ) to list external storage and navigate between them (but it won't work with USB)

2 - Using the native file selector (this will allow you to browse all files, including external services such as GoogleDrive, but does not allow you to customize the interface, customize the filter, or select files in different directories)

The first option seems better to me, but i am not sure about your priorities or if it could be enought for you. I wait for your feedback

**Comment by mul1sh** [comment]
@Maay Sorry got abit caught up :( so I forgot to push the PR but I've pushed it now. So kindly check on it and let me know what you think, I'll fix any issues you find if any in record speed this time 😃 thanks for the patience.

**Comment by mul1sh** [comment]
any comment on this?? thanks
